gaphor/storage/recovery.py

```diff
@@ -0,0 +1,441 @@
+            return
+        self.events.append(
+            (
+                "d",
+                event.element.id,
+                event.property.name,
+                event.old_value and event.old_value.id,
+            )
+        )
+
+    # We should handle all RevertibleEvent subtypes separately:
+    # gaphor.core.modeling.presentation.MatrixUpdated
+    # gaphor.diagram.presentation.HandlePositionEvent
+    # gaphor.diagram.connectors.ItemConnected
+    # gaphor.diagram.connectors.ItemDisconnected
+    # ignore: gaphor.diagram.connectors.ItemTemporaryDisconnected
+    # gaphor.diagram.connectors.ItemReconnected
+    # gaphor.diagram.tools.segment.LineSplitSegmentEvent
+    # gaphor.diagram.tools.segment.LineMergeSegmentEvent
+
+    @event_handler(MatrixUpdated)
+    def on_matrix_updated(self, event: MatrixUpdated):
+        self.events.append(("mu", event.element.id, tuple(event.element.matrix)))
+
+    @event_handler(HandlePositionEvent)
+    def on_handle_position_event(self, event: HandlePositionEvent):
+        pos = event.element.handles()[event.handle_index].pos
+        self.events.append(
+            ("hp", event.element.id, event.handle_index, (float(pos.x), float(pos.y)))
+        )
+
+    @event_handler(ItemConnected)
+    def on_item_connected(self, event: ItemConnected):
+        self.events.append(
+            (
+                "ic",
+                event.element.id,
+                event.handle_index,
+                event.connected_id,
+                event.port_index,
+            )
+        )
+
+    @event_handler(ItemDisconnected)
+    def on_item_disconnected(self, event: ItemDisconnected):
+        self.events.append(
+            (
+                "id",
+                event.element.id,
+                event.handle_index,
+                event.connected_id,
+                event.port_index,
+            )
+        )
+
+    @event_handler(ItemReconnected)
+    def on_item_reconnected(self, event: ItemReconnected):
+        self.events.append(
+            (
+                "ir",
+                event.element.id,
+                event.handle_index,
+                event.connected_id,
+                event.port_index,
+            )
+        )
+
+    @event_handler(LineSplitSegmentEvent)
+    def on_split_line_segment_event(self, event: LineSplitSegmentEvent):
+        self.events.append(
+            (
+                "ls",
+                event.element.id,
+                event.segment,
+                event.count,
+            )
+        )
+
+    @event_handler(LineMergeSegmentEvent)
+    def on_merge_line_segment_event(self, event: LineMergeSegmentEvent):
+        self.events.append(
+            (
+                "lm",
+                event.element.id,
+                event.segment,
+                event.count,
+            )
+        )
+
+
+def replay_events(events, element_factory, modeling_language):
+    for event in events:
+        match event:
```

Comment:
Nice usage of the match statement :star_struck:

Reply:
This is actually the first time I use this construct 😄